### PR TITLE
[MODTLR-42] Reorder secondary requests in sync with primary requests

### PR DESCRIPTION
**Pull request by roman-barannyk**
## Purpose
Handle Kafka event generated on request update.

When updated request is a primary request (ecsRequestPhase=Primary) and its position updated, secondary request should also be updated.

Load the queue of the primary request (TLR is enabled, it should be a unified queue - all requests with the same instanceId)

Find primary requests in this queue and get their corresponding secondary requests.

Load all request queues for all of the secondary requests (again, unified queues because TLR should always be enabled in data tenants in Phase I). There could be multiple such queues because secondary requests could be located in different tenants.

For each of the secondary request queues:

Check if secondary requests in the queue have the same order (not necessarily same positions) as their corresponding primary requests. If they do, skip this queue.

If they don’t, reorder secondary requests to match the order (not necessarily the positions) of their corresponding primary requests. One way to do this is to give them positions starting from 1 and then put all of the non-ECS requests at the end of the queue.

Update secondary requests with their new positions.

Note: add enough logging details when implementing items 1-5 to be able to close this story after checking the logs.

Resolves: [MODTLR-42](https://folio-org.atlassian.net/browse/MODTLR-42)

### CODE REVIEW

File: src/main/java/org/folio/service/impl/RequestEventHandler.java [hunk 1]
```diff
@@ -251,4 +324,31 @@ private static <T, V> boolean valueIsNotEqual(T o1, T o2, Function<T, V> valueEx
...
+    List<EcsTlrEntity> ecsTlrQueue) {
+
+    List<EcsTlrEntity> sortedEcsTlrQueue = new ArrayList<>(ecsTlrQueue);
+    Map<UUID, Integer> indexMap = new HashMap<>();
+    for (int i = 0; i < sortedPrimaryRequestIds.size(); i++) {
+      indexMap.put(sortedPrimaryRequestIds.get(i), i);
+    }
+
+    sortedEcsTlrQueue.sort(Comparator.comparingInt(entity -> indexMap.getOrDefault(
+      entity.getPrimaryRequestId(), Integer.MAX_VALUE)));
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/service/impl/RequestEventHandler.java:350]:
```suggestion

    Map<UUID, EcsTlrEntity> ecsTlrByPrimaryRequestId = ecsTlrQueue.stream()
        .collect(toMap(EcsTlrEntity::getPrimaryRequestId, Function.identity()));

    List<EcsTlrEntity> sortedEcsTlrs = sortedPrimaryRequestIds
      .stream()
      .map(ecsTlrByPrimaryRequestId::get)
      .toList();

```

File: src/main/java/org/folio/service/impl/RequestEventHandler.java [hunk 2]
```diff
@@ -224,6 +233,11 @@ private void propagateChangesFromPrimaryToSecondaryRequest(EcsTlrEntity ecsTlr,
       clonePickupServicePoint(ecsTlr, pickupServicePointId);
     }
 
+    if (valueIsNotEqual(primaryRequest, event.getData().getOldVersion(), Request::getPosition)) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/service/impl/RequestEventHandler.java:236]: move this into `handlePrimaryRequestUpdate`

File: src/main/java/org/folio/service/impl/RequestServiceImpl.java [hunk 3]
```diff
@@ -115,6 +116,11 @@ public Request updateRequestInStorage(Request request, String tenantId) {
       () -> requestStorageClient.updateRequest(request.getId(), request));
   }
 
+  @Override
+  public List<Request> getRequestsByInstanceId(String instanceId) {
+    return requestStorageClient.getRequestsByQuery(String.format("?query=instanceId==%s", instanceId));
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/service/impl/RequestServiceImpl.java:121]: add sorting to query?

File: src/main/java/org/folio/client/feign/RequestStorageClient.java [hunk 4]
```diff
@@ -17,4 +20,6 @@ public interface RequestStorageClient {
   @PutMapping("/{requestId}")
   Request updateRequest(@PathVariable String requestId, @RequestBody Request request);
 
+  @GetMapping(params = "query")
+  List<Request> getRequestsByQuery(@RequestParam("query") String query);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/client/feign/RequestStorageClient.java:24]: See if we can use Request Queue API in mod-circulation:
https://s3.amazonaws.com/foliodocs/api/mod-circulation/r/circulation.html#

File: src/main/java/org/folio/listener/kafka/KafkaEventListener.java [hunk 5]
```diff
@@ -53,6 +58,18 @@ public void handleRequestEvent(String eventString, MessageHeaders messageHeaders
     log.info("handleRequestEvent:: event consumed: {}", event::getId);
   }
 
+  @KafkaListener(
+    topicPattern = "${folio.environment}\\.\\w+\\.circulation\\.requests-batch-update",
```
Comment (alexanderkurash) [inline src/main/java/org/folio/listener/kafka/KafkaEventListener.java:62]: ${folio.environment}\\.\\w+\\.circulation\\.request-queue-reordering

**Review by alexanderkurash — APPROVED** [review]

**Review by OleksandrVidinieiev — APPROVED** [review]

**Pull request merged**